godot-core/src/builtin/plane.rs

```diff
@@ -0,0 +1,128 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use std::ops::Neg;
+
+use godot_ffi as sys;
+use sys::{ffi_methods, GodotFfi};
+
+use super::{is_equal_approx, real, Vector3};
+
+/// 3D plane in Hessian form: `a*b + b*y + c*z + d = 0`
+///
+/// Currently most methods are only available through [`InnerPlane`](super::inner::InnerPlane).
+///
+/// Note: almost all methods on `Plane` require that the `normal` vector have
+/// unit length and will panic if this invariant is violated. This is not separately
+/// annotated for each method.
+#[derive(Copy, Clone, PartialEq, Debug)]
+#[repr(C)]
+pub struct Plane {
+    pub normal: Vector3,
+    pub d: real,
+}
+
+impl Plane {
+    /// Creates a new `Plane` from the `normal` and the distance from the origin `d`.
+    ///
+    /// # Panics
+    /// In contrast to construction via `Plane { normal, d }`, this verifies that `normal` has unit length, and will
+    /// panic if this is not the case.
+    ///
+    /// _Godot equivalent: `Plane(Vector3 normal, float d)`_
+    #[inline]
+    pub fn new(normal: Vector3, d: real) -> Self {
+        let plane = Self { normal, d };
+        plane.assert_normalized();
+        plane
+    }
+
+    /// Create a new `Plane` through the origin from a normal.
+    ///
+    /// # Panics
+    /// See [`Self::new()`].
+    ///
+    /// _Godot equivalent: `Plane(Vector3 normal)`_
+    #[inline]
+    pub fn new_origin(normal: Vector3) -> Self {
+        Self::new(normal, 0.0)
+    }
+
+    /// Create a new `Plane` from normal and a point in the plane.
+    ///
+    /// # Panics
+    /// See [`Self::new()`].
+    ///
+    /// _Godot equivalent: `Plane(Vector3 normal, Vector3 point)`_
+    #[inline]
+    pub fn new_normal_point(normal: Vector3, point: Vector3) -> Self {
+        Self::new(normal, normal.dot(point))
+    }
+
+    /// Creates a new `Plane` from normal and origin distance.
+    ///
+    /// `a`, `b`, `c` are used for the `normal` vector.
+    /// `d` is the distance from the origin.
+    ///
+    /// # Panics
+    /// See [`Self::new()`].
+    ///
+    /// _Godot equivalent: `Plane(float a, float b, float c, float d)`_
+    #[inline]
+    pub fn from_coordinates(a: real, b: real, c: real, d: real) -> Self {
+        Self::new(Vector3::new(a, b, c), d)
+    }
+
+    /// Creates a new `Plane` from three points, given in clockwise order.
+    ///
+    /// If all three points are collinear, returns `None`.
+    ///
+    /// _Godot equivalent: `Plane(Vector3 point1, Vector3 point2, Vector3 point3)`_
+    #[inline]
+    pub fn from_points(a: Vector3, b: Vector3, c: Vector3) -> Option<Self> {
+        let normal = (a - c).cross(a - b).normalized();
```

Comment:
it is yeah, the cross product will simply be 0. either way we want to create a plane that is usable without crashing at the end here so we'd need to normalize the normal eventually. all the other constructors disallow creating a non-normalized one. 

as for why this one uses option and the others panic, i just followed what the gdnative implementation did.

Reply:
> either way we want to create a plane that is usable without crashing at the end here so we'd need to normalize the normal eventually

Yes, but then this could be done only in the `Some` case.

---

> as for why this one uses option and the others panic, i just followed what the gdnative implementation did.

Good point -- it probably makes sense to align the behavior.

I would probably tend to panic and add a `# Panic` section, because:
* the user most likely just uses `unwrap()` otherwise, which has a worse error message
* this class has just "weak invariants" due to public fields, so we should focus on detecting bugs and not get in the user's way
* it's always possible to add `try_from_points` later, if truly needed